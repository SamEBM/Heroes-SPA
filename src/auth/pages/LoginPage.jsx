import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {
    const navigate = useNavigate();
    const {login} = useContext(AuthContext);
    
    const onLogin = () => {
        login('Samuel Miranda');
        navigate('/', {replace: true});
    };

    return (
        <div className='container mt-5'>
            <h1>Login</h1>
            <p>This is a mock Login Page to test the use of private and public routes. To access the app simply click on the button.</p>
            <hr />

            <button onClick={onLogin} className='btn btn-primary'>
                Login
            </button>
        </div>
    )
};
