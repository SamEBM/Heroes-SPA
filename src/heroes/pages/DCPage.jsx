import React from 'react'

export const DCPage = () => {
    return (
        <h1 className='text-center mt-2'>DC Heroes</h1>
    )
}
