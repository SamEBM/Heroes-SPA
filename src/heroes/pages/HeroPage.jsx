import React from 'react'

export const HeroPage = () => {
    return (
        <h1 className='text-center mt-2'>Hero Page</h1>
    )
}
