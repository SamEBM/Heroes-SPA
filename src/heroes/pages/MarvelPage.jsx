import React from 'react'

export const MarvelPage = () => {
    return (
        <h1 className='text-center'>MarvelPage</h1>
    )
}
