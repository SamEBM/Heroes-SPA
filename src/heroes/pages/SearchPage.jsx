import React from 'react';
import { useForm } from '../../hooks/useForm';
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { HeroCard } from '../components';
import { getHeroesByName } from '../helpers';

export const SearchPage = () => {

    const navigate = useNavigate();
    const location = useLocation();
    const {q = ''} = queryString.parse(location.search);

    const heroes = getHeroesByName(q);

    const {searchHero, onInputChange} = useForm({
        searchHero: q,
    });

    const onSearchSubmit = (ev) => {
        ev.preventDefault();
        navigate(`?q=${searchHero}`);
    };

    return (
        <>
            <h1 className='text-center mt-2'>Search Heroes</h1>
            <hr />

            <div className="row">
                <div className='col-5'>
                    <h4>Who is your hero?</h4>
                    <hr />
                    <form onSubmit={onSearchSubmit}>
                        <input value={searchHero} onChange={onInputChange} className='form-control' name="searchHero" type="text" autoComplete="off" placeholder='Search a hero' />
                        <button className='btn btn-outline-primary mt-2'>Search</button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>

                    {
                        !q && 
                        <div className='alert alert-primary animate__animated animate__fadeIn'>Search a hero</div>
                    }

                    {
                        !heroes.length && q && 
                        <div className='alert alert-danger animate__animated animate__fadeIn'> Unable to find a hero with <b>{q}</b></div>
                    }

                    {
                        heroes.map( hero => (
                            <HeroCard key={hero.id} {...hero}/>
                        ))
                    }
                </div>
            </div>
        </>
    )
}
