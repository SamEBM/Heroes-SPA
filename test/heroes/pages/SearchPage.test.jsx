import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

// Para poder verificar que se ejecuta el navigate --------

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}));

// --------------------------------------------------------

describe('Pruebas en <SearchPage />', () => { 

    beforeEach(() => jest.clearAllMocks());

    test('debe de mostrarse correctamente con valores por defecto', () => {
        const {container} = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    test('debe de mostrar a Batman y el input con el valor del query string', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');
        
        const img = screen.getByRole('img');
        expect(img.src).toContain('batman');
    });

    test('debe de mostrar un error si no se encuentra el hero', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=antman']}>
                <SearchPage />
            </MemoryRouter>
        );

        const alert = screen.getByLabelText('no-hero');
        expect(alert).toBeTruthy();
        expect(screen.getByText('antman')).toBeTruthy();
    });

    test('debe llamar al navigate a una pantalla nueva', () => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {name: 'searchHero', value: 'superman'}});

        const form = screen.getByRole('form');
        fireEvent.submit(form);

        expect(mockUseNavigate).toHaveBeenCalledWith('?q=superman');
    });
});