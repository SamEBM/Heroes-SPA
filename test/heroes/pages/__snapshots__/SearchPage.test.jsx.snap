// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    class="text-center mt-2"
  >
    Search Heroes
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-12 col-md-5"
    >
      <h4>
        Who is your hero?
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchHero"
          placeholder="Enter the name of your hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-12 col-md-7 mt-5 mt-sm-5 mt-md-0"
    >
      <h4>
        Results
      </h4>
      <div
        aria-label="no-input"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
